src/main/java/com/ita/if103java/ims/service/impl/WarehouseItemAdviceServiceImpl.java

```diff
@@ -0,0 +1,109 @@
+                                          WarehouseAssociateDistanceService distanceService,
+                                          BestTradeService bestTradeService,
+                                          TopWarehouseDao topWarehouseDao,
+                                          AddressLinkerDao addressLinkerDao,
+                                          ItemService itemService,
+                                          WarehouseService warehouseService,
+                                          AssociateService associateService) {
+        this.calculationService = calculationService;
+        this.distanceService = distanceService;
+        this.bestTradeService = bestTradeService;
+        this.topWarehouseDao = topWarehouseDao;
+        this.addressLinkerDao = addressLinkerDao;
+        this.itemService = itemService;
+        this.warehouseService = warehouseService;
+        this.associateService = associateService;
+    }
+
+    @Override
+    public WarehouseItemAdviceDto getAdvice(Long itemId, Long accountId) {
+        final List<WeightAssociateDto> suppliers = bestTradeService.findBestSuppliersByItemId(itemId);
+        final List<WeightAssociateDto> clients = bestTradeService.findBestClientsByItemId(itemId);
+        final List<WarehouseToAssociateDistanceDto> warehouseAssociateDistances = keepIfAvailableRoute(
+            distanceService.getDistances(
+                addressLinkerDao.findWarehouseAddressesByIds(topWarehouseDao.findAllIds(accountId)),
+                addressLinkerDao.findAssociateAddressesByIds(getAssociateIds(suppliers)),
+                addressLinkerDao.findAssociateAddressesByIds(getAssociateIds(clients))
+            )
+        );
+        return toDto(itemId, suppliers, clients, calculationService.calculate(suppliers, clients, warehouseAssociateDistances));
+    }
+
+    private WarehouseItemAdviceDto toDto(Long itemId,
+                                         List<WeightAssociateDto> suppliers,
+                                         List<WeightAssociateDto> clients,
+                                         List<WarehouseIdAdviceDto> warehouseIdAdvices) {
+        return new WarehouseItemAdviceDto(
+            itemService.findById(itemId),
+            mapWarehouseAdvices(warehouseIdAdvices),
+            mapAssociates(suppliers),
+            mapAssociates(clients)
+        );
+    }
+
+    private List<AssociateDto> mapAssociates(List<WeightAssociateDto> associateDtoList) {
+        return associateDtoList.stream()
+            .map(WeightAssociateDto::getAssociateId)
+            .map(associateService::view)
+            .collect(Collectors.toUnmodifiableList());
+    }
+
+    private List<WarehouseAdviceDto> mapWarehouseAdvices(List<WarehouseIdAdviceDto> warehouseAdvices) {
+        return warehouseAdvices.stream()
+            .map(x -> new WarehouseAdviceDto(
+                warehouseService.findWarehouseById(x.getWarehouseId()),
```

Comment:
finding warehouses by id may be slow. From performance POV better to select all warehouse ids, query DB once, keep map of id and name and then in the loop get name from the map. 

Volodymyr did similar for events. But before change this, check if current performance is acceptable.

Reply:
ok, It's only quick temporary solution, I'll rewrite some queries to avoid those things